tokio/src/process/unix/orphan.rs

```diff
@@ -62,36 +54,87 @@ impl<T> OrphanQueueImpl<T> {
     fn len(&self) -> usize {
         self.queue.lock().unwrap().len()
     }
-}
 
-impl<T: Wait> OrphanQueue<T> for OrphanQueueImpl<T> {
-    fn push_orphan(&self, orphan: T) {
+    pub(crate) fn push_orphan(&self, orphan: T)
+    where
+        T: Wait,
+    {
         self.queue.lock().unwrap().push(orphan)
     }
-}
 
-impl<T: Wait> ReapOrphanQueue for OrphanQueueImpl<T> {
-    fn reap_orphans(&self) {
-        let mut queue = self.queue.lock().unwrap();
-        let queue = &mut *queue;
-
-        for i in (0..queue.len()).rev() {
-            match queue[i].try_wait() {
-                Ok(None) => {}
-                Ok(Some(_)) | Err(_) => {
-                    // The stdlib handles interruption errors (EINTR) when polling a child process.
-                    // All other errors represent invalid inputs or pids that have already been
-                    // reaped, so we can drop the orphan in case an error is raised.
-                    queue.swap_remove(i);
+    /// Attempts to reap every process in the queue, ignoring any errors and
+    /// enqueueing any orphans which have not yet exited.
+    pub(crate) fn reap_orphans(&self, handle: &SignalHandle)
+    where
+        T: Wait,
+    {
+        let has_changed = match self.sigchild.try_lock() {
+            Ok(mut guard) => match &mut *guard {
+                Some(sigchild) => sigchild.try_has_changed().and_then(Result::ok).is_some(),
```

Comment:
Maybe just move the `drain_orphan_queue` call up here? You only ever set `has_changed` to true in this branch. Also, maybe `try_has_changed` should just return a boolean?

Reply:
Originally the call to `drain_orphan_queue` was only done at the end of the method, but then I brought it up for the initialization case. I've compressed this without needing the temp boolean